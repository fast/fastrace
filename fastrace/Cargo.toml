[package]
name = "fastrace"
version = "0.7.7"

categories = ["development-tools::debugging"]
description = "A high-performance timeline tracing library for Rust"
documentation = "https://docs.rs/fastrace"
keywords = ["tracing", "span", "datadog", "jaeger", "opentelemetry"]
readme = "../README.md"

edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[features]
enable = []
serde = ["dep:serde"]

[dependencies]
fastant = "0.1"
fastrace-macro = { workspace = true }
parking_lot = "0.12"
pin-project = { version = "1.1.8" }
rand = "0.9"
rtrb = "0.3"
serde = { version = "1", optional = true }

[dev-dependencies]
async-trait = "0.1"
criterion = { version = "0.5", features = ["html_reports"] }
crossbeam = "0.8"
fastrace = { workspace = true, features = ["enable"] }
fastrace-datadog = { workspace = true }
fastrace-jaeger = { workspace = true }
fastrace-opentelemetry = { workspace = true }
flume = "0.11"
futures-timer = "3"
log = { workspace = true }
logcall = "0.1"
logforth = { version = "0.19.2", features = ["fastrace"] }
mockall = "0.13"
opentelemetry = { version = "0.28", features = ["trace"] }
opentelemetry-otlp = { version = "0.28", features = ["trace", "grpc-tonic"] }
opentelemetry_sdk = { version = "0.28", features = ["trace"] }
pollster = { version = "0.4.0" }
rand = "0.9"
rustracing = "0.6"
serial_test = "3.1"
test-harness = "0.3"
tokio = { version = "1.38", features = [
  "rt",
  "time",
  "macros",
  "rt-multi-thread",
] }
tracing = "0.1"
tracing-opentelemetry = "0.28.0"
tracing-subscriber = "0.3"

[[bench]]
harness = false
name = "trace"

[[bench]]
harness = false
name = "compare"

[[bench]]
harness = false
name = "spsc"

[[bench]]
harness = false
name = "object_pool"
