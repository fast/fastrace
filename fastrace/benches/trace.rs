// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use criterion::Criterion;
use criterion::black_box;
use criterion::criterion_group;
use criterion::criterion_main;
use fastrace::local::LocalCollector;
use fastrace::prelude::*;

fn init_fastrace() {
    struct DummyReporter;

    impl fastrace::collector::Reporter for DummyReporter {
        fn report(&mut self, _spans: Vec<fastrace::prelude::SpanRecord>) {}
    }

    fastrace::set_reporter(DummyReporter, fastrace::collector::Config::default());
}

fn dummy_iter(i: usize) {
    #[trace]
    fn dummy() {}

    for _ in 0..i {
        dummy();
    }
}

#[trace]
fn dummy_rec(i: usize) {
    if i > 1 {
        dummy_rec(i - 1);
    }
}

fn bench_trace_wide_raw(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_wide_raw");
    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let local_collector = LocalCollector::start();
                dummy_iter(*len);
                local_collector.collect()
            })
        });
    }

    group.finish();
    fastrace::flush();
}

fn bench_trace_wide(c: &mut Criterion) {
    init_fastrace();

    let mut group = c.benchmark_group("trace_wide");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let root = Span::root("root", SpanContext::new(TraceId(12), SpanId::default()));
                let _sg = root.set_local_parent();
                dummy_iter(*len - 1);
            })
        });
    }

    group.finish();
    fastrace::flush()
}

fn bench_trace_deep_raw(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_deep_raw");

    for len in &[1, 10, 100, 1000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let local_collector = LocalCollector::start();
                dummy_rec(*len);
                local_collector.collect()
            })
        });
    }

    group.finish();
    fastrace::flush()
}

fn bench_trace_deep(c: &mut Criterion) {
    init_fastrace();

    let mut group = c.benchmark_group("trace_deep");

    for len in &[1, 10, 100, 1000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let root = Span::root("root", SpanContext::new(TraceId(12), SpanId::default()));
                let _sg = root.set_local_parent();
                dummy_rec(*len - 1);
            })
        });
    }

    group.finish();
    fastrace::flush()
}

fn bench_trace_future(c: &mut Criterion) {
    init_fastrace();

    async fn f(i: u32) {
        for _ in 0..i - 1 {
            async {}.enter_on_poll(black_box("")).await
        }
    }

    let mut group = c.benchmark_group("trace_future");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let root = Span::root("root", SpanContext::new(TraceId(12), SpanId::default()));
                futures::executor::block_on(f(*len).in_span(root));
            })
        });
    }

    group.finish();
    fastrace::flush()
}

criterion_group!(
    benches,
    bench_trace_wide_raw,
    bench_trace_wide,
    bench_trace_deep_raw,
    bench_trace_deep,
    bench_trace_future
);
criterion_main!(benches);
