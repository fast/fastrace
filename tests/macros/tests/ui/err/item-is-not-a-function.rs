use fastrace::trace;

#[trace("test-span")]
struct S;

fn main() {}
