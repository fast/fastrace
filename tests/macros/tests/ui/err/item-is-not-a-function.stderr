error: expected identifier
 --> tests/ui/err/item-is-not-a-function.rs:5:9
  |
5 | #[trace("test-span")]
  |         ^^^^^^^^^^^
